File: openstacks/generate_problems.py
#! /usr/bin/env python

from __future__ import print_function

import math
import random


USAGE = """usage:
    %s MIN_SIZE MAX_SIZE STEP_SIZE MIN_ID MAX_ID DENSITY
where
    MIN_SIZE   the minimum problem size (number of products and orders)
    MAX_SIZE   the maximum problem size (number of products and orders)
    STEP_SIZE  the increment on problem size
    MIN_ID     the start number of problems of the same size
    MAX_ID     the end number of problems of the same size
    DENSITY    the approximate percentage of products included in each order"""


def main(args):
    if len(args) != 7:
        raise SystemExit(USAGE % args[0])
    min_size, max_size, step_size, min_id, max_id, density = list(map(int, args[1:]))
    for size in range(min_size, max_size + 1, step_size):
        for counter in range(min_id, max_id + 1):
            filename = "p%d_%d.txt" % (size, counter)
            generate_problem(filename, size, size, density)


def include_product_for_order(product, order, num_products, density):
    # This is based on Ioannis's code. It is designed so that products
    # are more likely to be included in orders if the distance between
    # the product and order id (before the reshuffling that happens at
    # the end) is small.
    #
    # Compared to an independent identically distributed choice for
    # each product/order pair, this should tend to favour instances
    # with small (= few stacks) solutions since it adds some locality
    # to the problem: the matrix entries (before shuffling) will tend
    # to cluster around the diagonal.

    d = abs(product - order)
    s = num_products / 6.0
    # The next line is the density function of the normal distribution
    # with mu = 0 and sigma = s for x = d.
    p = (1 / (s * (2 * math.pi) ** 0.5)) * math.exp(-(d * d) / (2 * s * s))
    # The next line is rather magical to me. If there is an
    # explanation for what this formula signifies, I don't know it.
    chance = p * density * num_products / (2.7 * (num_products / 5.0) ** 0.5)
    return random.randrange(100) <= chance


def generate_problem(filename, num_products, num_orders, density):
    products = list(range(num_products))
    orders = list(range(num_orders))
    matrix = [[0 for product in products] for order in orders]
    for order in orders:
        for product in products:
            if include_product_for_order(product, order, num_products, density):
                matrix[order][product] = 1
    for order in orders:
        if all(matrix[order][product] == 0 for product in products):
            matrix[order][random.choice(products)] = 1
    for product in products:
        if all(matrix[order][product] == 0 for order in orders):
            matrix[random.choice(orders)][product] = 1
    weights = [random.randrange(10, 100 + 1, 10) for product in products]

    random.shuffle(products)
    # Ioannis's original code, used for generating the competition
    # instances, only shuffles the orders but not the products, which
    # means that planners that process the products in numerical order
    # (or whatever) are somewhat biased. (Depending on the planner
    # and/or track, I guess this can be either positive or negative.)
    random.shuffle(orders)

    with open(filename, "w") as outfile:
        print(num_orders, num_products, file=outfile)
        for order in orders:
            row = [matrix[order][product] for product in products]
            print(" ".join(map(str, row)), file=outfile)
        print(" ".join(map(str, weights)), file=outfile)


if __name__ == "__main__":
    import sys

    main(sys.argv)

File: openstacks/generate_problems.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <string.h>

#define RANGE_MIN 0
#define RANGE_MAX 100

int A[300][300];

int generate_problem(int Products,int Orders,char *filename,int Density);

int main(int argc, char **argv)
{
	srand( (unsigned)time( NULL ) );
	if (argc!=7) {
		printf("Use the following parameters:\n");
		printf("MinItems, MaxItems, Step, MinID, MaxID, Density\n\nwhere:\n\n");
		printf("MinItems : the minimum number of Products=Orders per problem\n");
		printf("MaxItems : the minimum number of Products=Orders per problem\n");
		printf("Step : the increment on number of items\n");
		printf("MinID : the start number of number problems of the same size\n");
		printf("MaxID : the end number of number problems of the same size\n");
		printf("Density : the % probability that each product is included in each order\n");
		return -1;
	}
	int MinItems,MaxItems,Step,MinID,MaxID,Density;
	int counter;
	MinItems=atoi(argv[1]);
	MaxItems=atoi(argv[2]);
	Step=atoi(argv[3]);
	MinID=atoi(argv[4]);
	MaxID=atoi(argv[5]);
	Density=atoi(argv[6]);

	int Orders, Products;
	char filename[100];
	for (Products=MinItems;Products<=MaxItems;Products+=Step)
		for(counter=MinID;counter<=MaxID;counter++)
		{
			char temps[35];
			filename[0]='\0';
			strcat(filename, "p");
			sprintf(temps, "%d", Products);
			strcat(filename,temps);
			strcat(filename,"_");
			sprintf(temps, "%d", counter);
			strcat(filename,temps);
			strcat(filename,".txt");
			Orders=Products;
			generate_problem(Products,Orders,filename,Density);
		}

	return 0;
}

int compute_distance(int order, int product)
{
	int d;
	d=abs(order-product);
	return d;
}

int gaussian(int d, int Products, int Density)
{
	double s, p;
	s=(Products/6.0);
	p=(1/(s*sqrt(2*3.141)))*exp(-(d*d)/(2*s*s));
	return (int) ((double) p*Density*Products/(2.7*sqrt(Products/5.0)));
}

void swap_orders(int o1, int o2, int Orders, int Products)
{
	int Temp[300], i;
	if (o1==o2) return;
	for(i=0;i<Products;i++)
		Temp[i]=A[o1][i];
	for(i=0;i<Products;i++)
		A[o1][i]=A[o2][i];
	for(i=0;i<Products;i++)
		A[o2][i]=Temp[i];
}

int generate_problem(int Products,int Orders,char *filename,int Density){

	FILE *outfile;
	int i,j;
	int rand100;
	int count;

	outfile=fopen(filename,"w");
	fprintf(outfile,"%d %d\n",Orders,Products);

	for(i=0;i<Orders;i++)
		for(j=0;j<Products;j++)
		{
			int p;
			int D=compute_distance(i,j);
			rand100 = (int)(((double) rand() /
                         (double) RAND_MAX) * RANGE_MAX + RANGE_MIN);
//			if (rand100<=Density)
			p=gaussian(D,Products,Density);
			if (rand100<=p)
				A[i][j]=1;
			else
					A[i][j]=0;
		}

	// check for empty orders
	for(i=0;i<Orders;i++)
	{
		count=0;
		for(j=0;j<Products;j++)
			count+=A[i][j];
		if (count==0)
		{
			rand100 = (int)(((double) rand() /
                         (double) RAND_MAX) * (Products-1) );
			A[i][rand100]=1;
		}
	}

	// check for not used products
	for(j=0;j<Products;j++)
	{
		count=0;
		for(i=0;i<Orders;i++)
			count+=A[i][j];
		if (count==0)
		{
			rand100 = (int)(((double) rand() /
                         (double) RAND_MAX) * (Orders-1) );
			A[rand100][j]=1;
		}
	}

	// Reorder orders
	for(i=0;i<Orders;i++)
	{
		int o1, o2;
		o1= (int)(((double) rand() / (double) RAND_MAX) * (Orders-1) );
		o2= (int)(((double) rand() / (double) RAND_MAX) * (Orders-1) );
		swap_orders(o1,o2,Orders,Products);
	}

	// Writing to file...
	for(i=0;i<Orders;i++)
	{
		for(j=0;j<Products;j++)
			if (A[i][j]==1)
				fprintf(outfile,"1 ");
			else
				fprintf(outfile,"0 ");
		fprintf(outfile,"\n");
	}

	for (j=1;j<=Products;j++)
	{
		rand100 = 10*(int)(((double) rand() /
                         (double) RAND_MAX) * 10+1);
		fprintf(outfile,"%d ", rand100);
	}
	fprintf(outfile,"\n");

	fclose(outfile);
	return 1;
}
